plugin/SearchBarItem.qml

```diff
@@ -0,0 +1,243 @@
+** cancellation and also introduce a new iconized state.
+**
+** You may use this file under the terms of BSD license as follows:
+**
+** Redistribution and use in source and binary forms, with or without
+** modification, are permitted provided that the following conditions are met:
+**     * Redistributions of source code must retain the above copyright
+**       notice, this list of conditions and the following disclaimer.
+**     * Redistributions in binary form must reproduce the above copyright
+**       notice, this list of conditions and the following disclaimer in the
+**       documentation and/or other materials provided with the distribution.
+**     * Neither the name of the Jolla Ltd nor the
+**       names of its contributors may be used to endorse or promote products
+**       derived from this software without specific prior written permission.
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+** ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR
+** ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+****************************************************************************************/
+
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+
+BackgroundItem {
+    id: root
+
+    property alias iconized: searchField.iconized
+    property alias modelCount: searchField.modelCount
+    property real iconizedWidth
+
+    signal requestSearch(string text)
+    signal requestPreviousMatch()
+    signal requestNextMatch()
+    signal requestCancel()
+
+    onClicked: searchField.iconized = false
+
+    states: [State {
+                 name: "extended"
+                 when: !searchField.iconized
+                 PropertyChanges {
+                     target: searchField
+                     _margin: 0.
+                 }
+             },
+             State {
+                 name: "iconized"
+                 when: searchField.iconized
+                 PropertyChanges {
+                     target: root
+                     width: iconizedWidth
+                 }
+             }]
+    transitions: Transition {
+        NumberAnimation {
+            properties: "_margin, width"
+            easing.type: Easing.InOutQuad
+            duration: 400
+        }
+    }
+    Binding on highlighted {
+        target: root
+        when: searchIcon.down
+        value: true
+    }
```

Comment:
The reason is that both the BackgroundItem and the IconButton (the search icon) have a highlighted property that can be triggered independently. We want the IconButton to be highlighted when the BackgroundItem is down or when itself is down and vice versa. I was thinking that writing something like this would result in a loop:
<code>
BackgroundItem {
  id: root
  highlighted: down || icon.down
  IconButton {
    id: icon
    highlighted: down || root.down
  }
}
</code>
But well, now that I'm writing it (and not like <code>highlighted: down || other.highlighted</code> as firstly tried), You're right that it should be done like that and not with a Binding element.